doc/requirements/Requirements.md

```diff
@@ -0,0 +1,704 @@
+I.e.: a criteria **MUST** always have a state that can be matched by a rule.
+<why>if no rules can be formulated from a criterion, it is useless</why>
+
+Parameter values change **SHOULD** be selected by Rules.
+<why>A rule based inference engine has been chosen based on implementation and
+configuration ease</why>
+
+It **MUST** be possible to express a Rule that is always True.
+<why>In order to make a configuration "always applicable"</why>
+
+## Domains
+
+### Definition
+<dl>
+<dt>Domain</dt>
+<dd>
+Ordered set of configuration, all of which contain the values for the
+same parameters.
+</dd>
+</dl>
+
+### Requirement
+
+Each configuration **SHOULD** be in a "domain" (see Definition chapter).
+<why>Domains are mostly a way to define the priority of "Scenarios" for some
+parameters. It is not a MUST because this goal could also be achieve with (for
+example) global configurations and per parameter priority. It is not a MAY
+because the reference implementation uses domains.</why>
+
+If multiple configuration are eligible, the first one **MUST** be applied.
+<why>If multiple configuration are eligible</why>
+
+If no configuration is eligible, no configuration **MUST** be applied.
+<nb>It means that if none of the configurations is eligible, none is applied.
+This also mean that no function can be defined between criteria and states.
+I.e.: parameter values MAY depend on previous selection criterion states.</nb>
+<why>This is what the reference implementation does.</why>
+
+TODO sequence aware domains
+
+# (de)serialization
+
+## Philosophy
+Serialization and deserialization are meant to support destruction recovery and
+configuration deployment.
+
+These are the same requirements than for a database, it needs to be able to save
+its state and restore for backup, deployment, reboot...
+
+## Definition
+PF data includes:
+
+- parameters tree
+- configurations:
+    - selection rule
+    - parameter/value couples
+- domain:
+    - list of associated configurations
+    - order of priority
+
+## Requirement
+The PF data **MUST** be deserializable.
+<why>Otherwise a PF instance could only be created empty and then be filled by
+the tuning interface. The reference implementation supports it.</why>
+
+The PF data **SHOULD** be deserializable from a config file.
+<why>This is usually how program configuration are stored. The reference
+implementation supports it.</why>
+
+The PF data **SHOULD** be serializable.
+<why>In order to save a PF instance state and restore it later. This achieve
+destruction recovery. The reference implementation supports it.</why>
+
+The PF data **SHOULD** be serializable/deserializable by parts. <why>For easier
+configuration management: for versioning; for selecting only wanted parts of a
+complete configuration.</why>
+
+**TODO**: XML ?
+
+**TODO**: get the binary content of an element.
+
+# Introspection
+## Philosophy
+In order to debug
+
+## Requirements
+User **SHOULD** be able to inspect PF data.
```

Comment:
is that a reqirement of user or the lib implementer? you seem to talk about a debugging tool